### PR TITLE
[Feature](mlu-ops):Update the three-level flow template

### PR DESCRIPTION
**Pull request by mahxn0**
…abs operator support the stride feature

Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation
Update the three-level flow template

## 2. Modification

docs/MLU-OPS-OpList.md
modified:   docs/MLU-OPS-Operator-Development-Process.md
new file:   docs/design_docs/abs/abs.md
modified:   kernel_depends.toml
modified:   kernels/abs/abs.cpp
modified:   kernels/abs/abs.h
modified:   kernels/abs/abs_block.mlu
modified:   kernels/binary_op/binary_op_3pipeline.h
new file:   kernels/binary_op/binary_op_5pipeline.h
modified:   kernels/binary_op/binary_op_host.cpp
modified:   kernels/binary_op/binary_op_host.h
new file:   kernels/binary_op/binary_op_stride_3pipeline.h
new file:   kernels/binary_op/binary_op_tensor_scalar_3pipeline.h
modified:   kernels/div/div.cpp
modified:   kernels/div/div.h
deleted:    kernels/div/div_block.mlu
new file:   kernels/div/div_union1.mlu
modified:   kernels/log/log.cpp
modified:   kernels/log/log.h
deleted:    kernels/log/log_block.mlu
new file:   kernels/log/log_union1.mlu
modified:   kernels/sqrt/sqrt.cpp
modified:   kernels/sqrt/sqrt.h
deleted:    kernels/sqrt/sqrt_block.mlu
new file:   kernels/sqrt/sqrt_union1.mlu
new file:   kernels/unary_op/complex_unary_op_3pipeline.h
new file:   kernels/unary_op/complex_unary_op_stride_3pipeline.h
modified:   kernels/unary_op/unary_op_3pipeline.h
new file:   kernels/unary_op/unary_op_4pipeline.h
modified:   kernels/unary_op/unary_op_5pipeline.h
modified:   kernels/unary_op/unary_op_host.cpp
modified:   kernels/unary_op/unary_op_host.h
new file:   kernels/unary_op/unary_op_no_pipeline.h
new file:   kernels/unary_op/unary_op_stride_3pipeline.h
modified:   kernels/utils/common.h
modified:   test/mlu_op_gtest/pb_gtest/src/zoo/abs/abs.cpp
new file:   test/mlu_op_gtest/pb_gtest/src/zoo/abs/test_case/case_1_with_stride.prototxt
new file:   test/mlu_op_gtest/pb_gtest/src/zoo/abs/test_case/case_2_with_stride.prototxt
new file:   test/mlu_op_gtest/pb_gtest/src/zoo/abs/test_case/case_3_with_stride.prototxt
modified:   test/mlu_op_gtest/pb_gtest/src/zoo/div/div.cpp
modified:   test/mlu_op_gtest/pb_gtest/src/zoo/log/log.cpp
modified:   test/mlu_op_gtest/pb_gtest/src/zoo/sqrt/sqrt.cpp
modified:   test/mlu_op_gtest/pb_gtest/src/zoo/sqrt_backward/sqrt_backward.cpp

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/MLU-OPS-Operator-Development-Process.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ MLU-OPS™ 是面向 MLU 平台的人工智能网络加速库，算子的功能
 
 ### 文档开发阶段
 
+#### 算子设计文档
```
Comment (PetrelYy) [inline docs/MLU-OPS-Operator-Development-Process.md:37]: 章节名称一起统一了？（有两套一级标题），此外也 2.1 重复了

File: kernels/abs/abs_block.mlu [hunk 2]
```diff
@@ -25,84 +25,127 @@
...
+    size_t element_num) {
   if (d_type == MLUOP_DTYPE_FLOAT) {
+    KERNEL_CHECK(MLUBlockKernel3StagePipelineWithStrideAbsFloat<float, float>
+                 <<<k_dim, k_type, queue>>>((char *)x, x_shape, (char *)y,
+                                            y_shape, element_num));
+  } else if (d_type == MLUOP_DTYPE_HALF || d_type == MLUOP_DTYPE_BFLOAT16) {
     KERNEL_CHECK(
-        MLUBlockKernel5StagePipelineAbsfloatFast<<<k_dim, k_type, queue>>>(
-            (void *)x, (void *)y, num, 0.0));
+        MLUBlockKernel3StagePipelineWithStrideAbsHalfBfloat16<half, half>
```
Comment (PetrelYy) [inline kernels/abs/abs_block.mlu:140]: 这里是half，bf16 类型如何计算？
还是说把bf16 用half函数来做（第一位符号位）？那这样看，int 和float 实例化float就行？

**Review by DanieeelLiu — APPROVED** [review]

File: kernels/binary_op/binary_op_3pipeline.h [hunk 3]
```diff
@@ -24,141 +24,273 @@
...
-    __memcpy_async(nram_x + pong_x, base_addr_x + nram_limit, span_handle_size,
-                   GDRAM2NRAM);
-    __memcpy_async(nram_y + pong_y, base_addr_y + nram_limit, span_handle_size,
-                   GDRAM2NRAM);
-    // C
-    OpFunc(nram_x, nram_y, nram_aux1, nram_aux2, nram_aux3, nram_limit,
-           nram_limit);
-    __sync();
-  }
+#define BINARY_OP_KERNEL_3PIPELINE_DECLARE(Op, Prefer)                  \
```
Comment (PetrelYy) [inline kernels/binary_op/binary_op_3pipeline.h:34]: nl prefer 是不是删掉了？

File: kernels/binary_op/binary_op_5pipeline.h [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+      __sync_cluster();                                                        \
+    }                                                                          \
+                                                                               \
+    for (int32_t i = 0; i < repeat - 2; i++) {                                 \
+      __memcpy_async(base_cluster_out + i * sram_num,                          \
+                     sram_in1 + (i % 2) * sram_pong_gap,                       \
+                     sram_num * sizeof(DType_out), SRAM2GDRAM);                \
+      __memcpy_async(sram_in1 + (i % 2) * sram_pong_gap,                       \
+                     base_cluster_in1 + (i + 2) * sram_num,                    \
+                     sram_num * sizeof(DType_in1), GDRAM2SRAM);                \
```
Comment (PetrelYy) [inline kernels/binary_op/binary_op_5pipeline.h:120]: - S2G, G2S 同一块内存的读写串行
- nram_in1，S2N、计算、N2S 串行

只做了 G2S/S2G    与   S2N/N2S 的掩盖。 与owner确认下这套模板的偏向场景（IO瓶颈？）
， 此外这个模板没算子使用到，覆盖率上不去，有啥想法不？

File: kernels/sqrt/sqrt_union1.mlu [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+                 <<<k_dim, k_type, queue>>>((char *)x, (char *)y, num));
+  } else {
+    // bfloat16
+    KERNEL_CHECK(MLUBlockKernel3StagePipelineSqrtFast<bfloat16_t, bfloat16_t>
+                 <<<k_dim, k_type, queue>>>((char *)x, (char *)y, num));
+  }
+  return MLUOP_STATUS_SUCCESS;
+}
+
+mluOpStatus_t MLUOP_WIN_API Kernel3StagePipelineSqrtBackward(
```
Comment (PetrelYy) [inline kernels/sqrt/sqrt_union1.mlu:270]: abs 等算子新增了类型，未修改 mlu op.h

File: kernels/abs/abs_block.mlu (hunk 2 above)
Comment (mahxn0) [reply kernels/abs/abs_block.mlu:140]: 是的  bf16 用half函数来做，第一位符号位
int 和float 实例化float 也放一起了。

File: kernels/binary_op/binary_op_3pipeline.h (hunk 3 above)
Comment (mahxn0) [reply kernels/binary_op/binary_op_3pipeline.h:34]: 没有

**Review by PetrelYy — APPROVED** [review]

**Review by AndyQiao0828 — APPROVED** [review]

**Pull request merged**